Simplify polonius location-sensitive analysis

lqd opened the pull request:
This PR reworks the location-sensitive analysis into what we think is a worthwhile subset of the datalog analysis. A sort of polonius alpha analysis that handles NLL problem case 3 and more, but is still using the faster "reachability as an approximation of liveness", as well as the same loans-in-scope computation as NLLs -- and thus doesn't handle full flow-sensitivity like the datalog implementation.

In the last few months, we've identified this subset as being actionable:
- we believe we can make a stabilizable version of this analysis
- it is an improvement over the status quo
- it can also be modeled in a-mir-formality, or some other formalism, for assurances about soundness, and I believe @nikomatsakis is interested in looking into this during H2.
- and we've identified the areas of work we wish to explore later to gradually expand the supported cases: the differences between reachability and liveness, support of kills, and considerations of time-traveling, for example.

The approach in this PR is to try less to have the graph only represent live paths, by checking whether we reach a live region during traversal and recording the loan as live there, instead of equating traversal with liveness like today because it has subtleties with the typeck edges in statements (that could forward loans to the successor point without ensuring their liveness). We can then also simplify these typeck stmt edges. And we also can simplify traversal by removing looking at kills, because that's enough to handle a bunch of NLL problem 3 cases -- and we can gradually support them more and more in traversal in the future, to reduce the approximation of liveness.

There's still some in-progress pieces of work w/r/t opaque types that I'm expecting [lcnr's opaque types rework](https://github.com/rust-lang/rust/pull/139587), and [amanda's SCCs rework](https://github.com/rust-lang/rust/pull/130227) to handle. That didn't seem to show up in tests until I rebased today (and shows lack of test coverage once again) when https://github.com/rust-lang/rust/pull/142255 introduced a couple of test failures with the new captures rules from edition 2024. It's not unexpected since we know more work is needed with member constraints (and we're not even using SCCs in this prototype yet)

I'll look into these anyways, both for future work, and checking how these other 2 PRs would change things.

---

I'm not sure the following means a lot until we have some formalism in-place, but:
- I've changed the polonius compare-mode to use this analysis: the tests pass with it, except 2 cases with minor diagnostics differences, and the 2 edition 2024 opaque types one I mentioned above and need to investigate
- things that are expected to work still do work: it bootstraps, can run our rustc-perf benchmarks (and the results are not even that bad), and a crater run didn't find any regressions (forgetting that crater currently fails to test around a quarter of all crates 👼)
- I've added tests with improvements, like the NLL problem case 3 and others, as well as some that behave the same as NLLs today and are thus worse than the datalog implementation

r? @jackh726 

(no rush I know you're deep in phd work and "implmentating" the new trait solver for r-a :p <3)

This also fixes rust-lang/rust#135646, a diagnostics ICE from the previous implementation.

lqd committed:
switch polonius compare-mode to polonius=next

lqd committed:
update test expectations for boring locals + dropckoutlives interactions

The suboptimal error only appears with NLLs due to liveness differences
where polonius cannot have as many boring locals. Sometimes this causes
NLLs to emit a duplicate error as well.

lqd committed:
new impl fixes crash test

lqd committed:
turn expensive assert into debug assertion

lqd committed:
add some test cases for overlapping yielded items

These are just some sanity checks to ensure NLLs, the polonius alpha
analysis, and the datalog implementation behave the same on these common
examples.

lqd committed:
add NLL-like imprecision example

This test showcases the same imprecision as NLLs, unlike the datalog
implementation, when using reachability as a liveness approximation.

lqd committed:
add cursor-like example that works

This is an example similar to the linked-list cursor examples
where the alpha shows the same imprecision as NLLs, but that can work due to
the loans not being live after the loop, or the constraint graph being
simple enough that the cfg/subset relationships are the same for
reachability and liveness.

bors commented:
:umbrella: The latest upstream changes (presumably #142030) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

jackh726 reviewed (changes requested):
First thought: there are a bunch of really useful tests added here. I'd really like to see them added separately, covering nll, polonius-legacy, and polonius-next. You can use known-bug as appropriate.

lqd commented:
Could you expand on how useful having known-bug tests is for things that we don’t support yet and have open issues?

I guess you mean in another PR with "separately"?

jackh726 commented on tests/ui/nll/issue-46589.rs:
```diff
@@ -1,6 +1,6 @@
 //@ ignore-compare-mode-polonius (explicit revisions)
 //@ revisions: nll polonius_next polonius
-//@ [polonius_next] check-pass
+//@ [polonius_next] check-fail
```
I guess it's really a question of what's in scope for `known-bug` 😛 For things like polonius, where it's like "it works exactly like it's intended, but we want it to not work this way", it's on the edge. In my mind, if there's an issue for it, then it's `known-bug`.

lqd replied:
Done in this PR.

lqd committed:
mark polonius=next's NLL imprecisions as known-bugs

- linked-list cursor-like patterns
- issue-46589

These are known-bugs for the polonius alpha, where they show the same
imprecision as NLLs, but are supported by the old datalog
implementation.

lqd committed:
simplify polonius=next

Remove incomplete handling of kills during traversal for loan liveness
to get to a simpler and actionable prototype.

This handles the cases, on sufficiently simple examples, that were
deferred from NLLs (NLL problem case 3, lending iterators), and is still
a good step to put in people's hands without needing to wait for another
full implementation. This is a practical cut in scope, but it also
shows where are the areas of improvement, that we will explore in the
future.

lqd commented:
Rebased to remove and update the tests now that #145053 has landed.

jackh726 commented:
Seems good to me.

@bors r+

bors commented:
:pushpin: Commit b99fe2b72022f4fe7a6f2e55fd70cc8530566473 has been approved by `jackh726`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=jackh726 b99fe2b72022f4fe7a6f2e55fd70cc8530566473 -->
<!-- homu: {"type":"Approved","sha":"b99fe2b72022f4fe7a6f2e55fd70cc8530566473","approver":"jackh726","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->
